src/Polysemy/Internal/Union.hs

```diff
@@ -131,102 +138,130 @@ type Member e r = MemberNoError e r
+--
+-- @since 1.2.3.0
 type MemberWithError e r =
-  ( MemberNoError e r
+  ( Find r e
 #ifndef NO_ERROR_MESSAGES
     -- NOTE: The plugin explicitly pattern matches on
-    -- `WhenStuck (IndexOf r _) _`, so if you change this, make sure to change
+    -- `WhenStuck (LocateEffect r _) _`, so if you change this, make sure to change
     -- the corresponding implementation in
     -- Polysemy.Plugin.Fundep.solveBogusError
-  , WhenStuck (IndexOf r (Found r e)) (AmbiguousSend r e)
+  , LocateEffect r e ~ '()
+  , WhenStuck (LocateEffect r e) (AmbiguousSend r e)
 #endif
   )
 
 type MemberNoError e r =
   ( Find r e
-  , e ~ IndexOf r (Found r e)
+#ifndef NO_ERROR_MESSAGES
+  , LocateEffect r e ~ '()
+#endif
   )
 
-
 ------------------------------------------------------------------------------
--- | The kind of type-level natural numbers.
-data Nat = Z | S Nat
-
-
-------------------------------------------------------------------------------
--- | A singleton for 'Nat'.
-data SNat :: Nat -> Type where
-  SZ :: SNat 'Z
-  SS :: SNat n -> SNat ('S n)
-
-instance TestEquality SNat where
-  testEquality SZ     SZ     = Just Refl
-  testEquality (SS _) SZ     = Nothing
-  testEquality SZ     (SS _) = Nothing
-  testEquality (SS n) (SS m) =
-    case testEquality n m of
-      Nothing -> Nothing
-      Just Refl -> Just Refl
-  {-# INLINE testEquality #-}
-
-
-type family IndexOf (ts :: [k]) (n :: Nat) :: k where
-  IndexOf (k ': ks) 'Z = k
-  IndexOf (k ': ks) ('S n) = IndexOf ks n
-
-
-type family Found (ts :: [k]) (t :: k) :: Nat where
+-- | A proof that @e@ is an element of @r@.
+--
+-- Due to technical reasons, @'ElemOf' r e@ is not powerful enough to
+-- prove @'Member' e r@; however, it can still be used send actions of @e@
+-- into @r@ by using 'Polysemy.Internal.subsumeUsing'.
+data ElemOf r e where
+  Here :: ElemOf (e ': r) e
+  In   :: ElemOf r e -> ElemOf (e' ': r) e
+
+instance TestEquality (ElemOf r) where
+  testEquality Here Here
+    = Just Refl
+  testEquality (In e) (In e')
+    = testEquality e e'
+  testEquality _ _ =
+    Nothing
+
+type family LocateEffect (ts :: [k]) (t :: k) :: () where
 #ifndef NO_ERROR_MESSAGES
-  Found '[]       t = UnhandledEffect t
+  LocateEffect '[] t = UnhandledEffect t
 #endif
-  Found (t ': ts) t = 'Z
-  Found (u ': ts) t = 'S (Found ts t)
-
+  LocateEffect (t ': ts) t = '()
+  LocateEffect (u ': ts) t = LocateEffect ts t
 
 class Find (r :: [k]) (t :: k) where
-  finder :: SNat (Found r t)
+  membership' :: ElemOf r t
 
 instance {-# OVERLAPPING #-} Find (t ': z) t where
-  finder = SZ
-  {-# INLINE finder #-}
+  membership' = Here
+  {-# INLINE membership' #-}
+
+instance Find z t => Find (_1 ': z) t where
+  membership' = In $ membership' @_ @z @t
+  {-# INLINE membership' #-}
+
+------------------------------------------------------------------------------
+-- | A class for effect rows whose elements are inspectable.
+--
+-- This constraint is eventually satisfied as @r@ is instantied to a
```

Comment:
This comment suggests that this thing can never get stuck. Is that true? If so, how does it work?

Reply:
The comment is just so newcomers won't get scared about "what do I need to have `KnownRow` be satisfied?"; the instances defined take care of it eventually.

It won't always work out so cleanly. Given a type variable `s`, in order for `State s ': r` to be `KnownRow` you need to specify that `s` is a `Typeable`.